### PR TITLE
Fix for using std:: in the dual numbers

### PR DESCRIPTION
**Pull request by bslazarov**
Fix for compiling the current AD implementation for some of the std math functions with gcc. 

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3600](https://github.com/mfem/mfem/pull/3600) | @bslazarov | @v-dobrev | @kmittal2 + @jandrej | 5/5/23 | 9/15/23 | 9/25/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by v-dobrev** [comment]
What version of gcc needs these changes? What are the error messages?

**Comment by bslazarov** [comment]
> What version of gcc needs these changes? What are the error messages?

/mnt/home/develop/wrk/mtopel/mfem/miniapps/autodiff/../../linalg/dual.hpp:370:26: error: no matching function for call to ‘pow(double&, double&)’
  370 |    value_type value = pow(a.value, b.value);
      |                       ~~~^~~~~~~~~~~~~~~~~~
/mnt/home/develop/wrk/mtopel/mfem/miniapps/autodiff/../../linalg/dual.hpp:366:33: note: candidate: ‘template<class value_type, class gradient_type> mfem::internal::dual<value_type, gradient_type> mfem::internal::pow(mfem::internal::dual<value_type, gradient_type>, mfem::internal::dual<value_type, gradient_type>)’
  366 | dual<value_type, gradient_type> pow(dual<value_type, gradient_type> a,
      |                                 ^~~
/mnt/home/develop/wrk/mtopel/mfem/miniapps/autodiff/../../linalg/dual.hpp:366:33: note:   template argument deduction/substitution failed:
/mnt/home/develop/wrk/mtopel/mfem/miniapps/autodiff/../../linalg/dual.hpp:370:26: note:   mismatched types ‘mfem::internal::dual<value_type, gradient_type>’ and ‘double’
  370 |    value_type value = pow(a.value, b.value);

version is: gcc version 9.4.0

**Comment by bslazarov** [comment]
> What version of gcc needs these changes? What are the error messages?

here comes the example:
#include "mfem.hpp"
#include "../../linalg/dual.hpp"

template<class DType>
DType test_f(DType* inp){
    //DType rez={double(0.0),double(1.0)};
    DType rez={0.0};
    //DType
    rez=pow(inp[0],inp[1]);
    //DType rez=pow(5.0,inp[1]);
    return rez;
}


int main(int argc, char *argv[])
{

    typedef mfem::internal::dual<double,double> ADType;

    ADType inp[2];
    inp[0].value=2.0; inp[0].gradient=0.0;
    inp[1].value=1.4; inp[1].gradient=1.0;

    ADType rez;
    rez=test_f<ADType>(inp);

    std::cout<<"grad_1="<<rez.gradient<<std::endl;

}

**Comment by v-dobrev** [comment]
Looking at the errors, they seem to be only in places where the function call does not use the fully qualified `std::pow` but just the short version, `pow`. Can't you just add the `std::` prefix to fix this?

**Comment by bslazarov** [comment]
> Looking at the errors, they seem to be only in places where the function call does not use the fully qualified `std::pow` but just the short version, `pow`. Can't you just add the `std::` prefix to fix this?

std::pow is not defined for  value_type=dual<double,double>, and thus, will not work in the general case.

File: linalg/dual.hpp [hunk 1]
```diff
@@ -342,48 +349,60 @@ dual<value_type, gradient_type> tan(dual<value_type, gradient_type> a)
...
 }
 
 /** @brief implementation of `a` (dual) raised to the `b` (dual) power */
 template <typename value_type, typename gradient_type> MFEM_HOST_DEVICE
 dual<value_type, gradient_type> pow(dual<value_type, gradient_type> a,
                                     dual<value_type, gradient_type> b)
 {
+   using std::log;
+   using std::pow;
    value_type value = pow(a.value, b.value);
```
Comment (v-dobrev) [inline linalg/dual.hpp:379]: I'm talking about lines like this. Here, you only need `pow(double,double)`, right?

Comment (v-dobrev) [reply linalg/dual.hpp:379]: Or, do you actually need nested dual types?

Comment (bslazarov) [reply linalg/dual.hpp:379]: I do need it for nested dual types - the nesting is used for second derivatives and is a functionality which I need.

Comment (v-dobrev) [reply linalg/dual.hpp:379]: Oh, okay, why didn't you say so earlier? 😄

This means you want this `pow` call to resolve to `std::pow` when using it with `value_type=double`, and to `mfem::internal::pow` when using it with `value_type=dual<double,double>`.

By the way, this fix seems to be necessary with Clang as well, not just with GCC, even without the nested dual types.

**Comment by v-dobrev** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Review by kmittal2 — APPROVED** [review]

**Comment by v-dobrev** [comment]
ping: @jandrej

**Review by jandrej — APPROVED** [review]

**Comment by v-dobrev** [comment]
Merged in `next` for testing...

**Pull request merged**